stonesoup/updater/pointprocess.py

```diff
@@ -145,6 +145,136 @@ def _calculate_update_terms(updated_sum_list, hypotheses):
         return 1
 
 
+class SoftProcessUpdater(Base):
+    r"""
+    Base updater class for the implementation of PHD-EF filter.
+    """
+    updater = Property(
+        KalmanUpdater,
+        doc="Underlying updater used to perform the \
+             single target Kalman Update.")
+
+    clutter_spatial_density = Property(
+        float,
+        default=1e-26,
+        doc="Spatial density of the clutter process uniformly\
+             distributed across the state space.")
+
+    normalisation = Property(
+        bool,
+        default=True,
+        doc="Flag for normalisation")
+
+    prob_detection = Property(
+        Probability,
+        default=1,
+        doc="Probability of a target being detected at the current timestep")
+
+    prob_survival = Property(
+        Probability,
+        default=1,
+        doc="Probability of a target surviving until the next timestep")
+
+    def update(self, hypotheses):
+        """
+        Updates the current components in a
+        :class:`GaussianMixture` by applying the underlying \
+        :class:`KalmanUpdater` updater to each component \
+        with the supplied measurements.
+
+        Parameters
+        ==========
+        hypotheses : list of :class:`MultipleHypothesis`
+            Measurements obtained at time :math:`k+1`
+
+        Returns
+        =======
+        updated_components : :class:`GaussianMixtureUpdate`
+            GaussianMixtureMultiTargetTracker with updated \
+            components at time :math:`k+1`
+        """
+        updated_components = list()
+        weight_sum_list = list()        
+        single_hypothesis_list = list()
+        number_detections = set()
+        for multi_hypothesis in hypotheses[:-1]:
+            for hypothesis in multi_hypothesis:
+                single_hypothesis_list.append(hypothesis)
+                number_detections.add(hypothesis.measurement[0].tag)
+        # Loop over all GM components of soft measurements 
+        for num_detect in number_detections:
+            updated_measurement_components = list()
+            # Initialise weight sum for measurement to clutter intensity
+            weight_sum = 0
+            # For every valid single hypothesis, update that GM component with
+            # measurements and calculate new weight
+            for hypothesis in single_hypothesis_list:
+                if (hypothesis.measurement.components[0].tag == num_detect):
```

Comment:
As you mentioned, Step 3 of the Algorithm has 3 recursions (for loops).  I have attached a file, in which more details is provided.

Each soft measurement is modeled as the summation of weighted Gaussian components (GM components). For example, following soft_measurement has 4 components:
GaussianMixtureDetection([
                    TaggedWeightedGaussianState(([0, 0]), np.diag([20, 20]), timestamp=time, weight=0.3, tag=11),
                    TaggedWeightedGaussianState(([1, 1]), np.diag([30, 30]), timestamp=time, weight=0.1, tag=11),
                    TaggedWeightedGaussianState((2, 2]), np.diag([40, 40]), timestamp=time, weight=0.2, tag=11),
                    TaggedWeightedGaussianState(([3, 3]), np.diag([50, 50]), timestamp=time, weight=0.4, tag=11)],
                    timestamp=time, measurement_model=cartesian_measurement_model), 
It is seen that each GM component has (state_vector, covariance_matrix, timestamp, weight, tag). Notice that the summation of weights should be equal to 1 (Here, 0.3+0.1+0.2+0.4=1). Also, tag of all GM components of eah soft_measurement has the same value (Here 11) (In this way we understand these GM components belongs to the same soft_measurement).

In hypothesis\distance.py we create one SingleDistanceHypothesis for each GM components of each soft measurement.  Therefore, here we do not have an additional recursion you mentioned above. 
Using "if (hypothesis.measurement.components[0].tag == num_soft_detect):" enables us to iterate only over GM components of the same soft_measurements since their tag is unique and same. We need to do this because of the required Normalization shown in the attached file.  

So, first recursion is over different measurements (Same as GM-PHD filter). Second recursion is over GM components of the same soft_measurement (GM-PHD doen't have this recursion). Third recursion is same as GM-PHD filter.

Reply:
![image](https://user-images.githubusercontent.com/53274822/96608364-4b891900-12c7-11eb-9a1b-25c73836d319.png)